### ISSUE TITLE
`setItems` does not work in the bulk mode (after `beginUpdate(true)`)

### ISSUE THREAD

**Issue by Mingun** [6pac/SlickGrid#621]
Calling `ensureIdUniqueness()` in the bulk mode update causes throwing an error:
```js
data.beginUpdate(true);
data.setItems(json);// throws "Error: [SlickGrid DataView] Each data element must implement a unique 'id' property"
data.endUpdate();
```
This is happened because in the bulk mode `idxById` map don't filled:
https://github.com/6pac/SlickGrid/blob/6495b69bdb701466507fb8d7e9fec089bdd60bc9/slick.dataview.js#L171-L175
but this map is checked in the `ensureIdUniqueness`:
https://github.com/6pac/SlickGrid/blob/6495b69bdb701466507fb8d7e9fec089bdd60bc9/slick.dataview.js#L186-L194
and check is performed just after creation of empty `idxById`, so `idxById.get(id) !== i` always `true`:
https://github.com/6pac/SlickGrid/blob/6495b69bdb701466507fb8d7e9fec089bdd60bc9/slick.dataview.js#L204-L214

**Comment by ghiscoding** [comment]
It works ok for me and that code was there since the beginning with a slight change, it used to be `idxById[id] !== i` instead of the new `idxById.get(id) !== i`. However compare to before the bulk code got implemented, the Id that you provide were not strictly check on equality but since the new bulk code it is now it requires a strictly equality check, in other words (an id of "1" == 1 was working before the bulk implementation but it no longer works and I would bet that is your issue so check your Id type are the same string or number but they must match on both side).

**Comment by ghiscoding** [comment]
Well like I said, I don't have that problem so you seem to be the only person with that problem (apart from another user who mentioned the small breaking change that I mentioned in my previous post about the strictly type check). `Slick.Map` is just a wrapper that will point to [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) for new browsers or will use a polyfill for old browser like IE and Map does a strictly equal check. Apart from the Map, I don't think there's anything new, the `idxById.get(id) !== i` has been in the code for the past 10 years except that it used to be `idxById[id] !== i`

I'm not sure if @Danielku15 can add more details (he's the person who implemented that Bulk feature), I can't help you more than that, sorry

**Comment by Mingun** [comment]
I've created a JSFiddle snippet with demonstration: https://jsfiddle.net/eur7jbfa/
![setItems-bug](https://user-images.githubusercontent.com/450131/125480706-05a39fa3-cdbd-4f50-a995-594099d0a0b2.png)

**Comment by Danielku15** [comment]
I would have to dive again into the details of the bulk logic (already has been a while since I added it). In general the idea behind the new bulk mode is to have optimized modifications. `setItems` is a full replacement of all items in the list, I would need to go through the code flow, but is there a particular point why you are trying to use `beginUpdate(true)` over the default `beginUpdate()`?

It is not nice that it actually fails with an error like this and it should be possible to handle the logic correctly as it is almost equal to a delete all and add new. 

But I think there will not be any significant impact between `beginUpdate(true)` and `beginUpdate()` (or with false) because the magic of the new bulk is in inserting/deleting/adding new items. It delays certain id lookup updates to the `endUpdate` which can be costly. On a full replace, this has not much meaning. As a temporary workaround I would recommend to just use the normal `beginUpdate()` not the bulk variant. 

As a full fix I can think of following approaches: 

1. Throw an appropriate error to the user that setItems is not available during bulk update (and document it). 
2. fully remove the `ensureIdUniqueness` from the code and by this avoid the error. Even in the old implementation [before my bulk system](https://github.com/6pac/SlickGrid/blob/32e03d5a04b576e9e985b6a3321b82eb97ace3a9/slick.dataview.js) this uniqueness check only was done on setItems and not on any modifications later on. Therefore it does not have much value unless we add the right logic everywhere. 
3. Suspend the ID uniqueness check for bulk operations and delay it to the `endUpdate`. This can have unexpected side effects because the uniqueness check did not exist before on modifications as described in point 2. 


Long story short: I recommend to use the non-bulk update as workaround and would go for option 1. It's the simplest solution to inform users that this operation is not supported in this mode. As it does not make much sense anyhow it is not a big loss.

**Comment by 6pac** [comment]
@mingun this issue should be fixed in the next release (2.4.39)

**Comment by Mingun** [comment]
Thanks! Actually, I don't quite understand the problem. `beginUpdate()/endUpdate()` calls _already_ implements bulk change mode. You simple don't need to have this functions if you update view on each data change because you can just initiate redraw in `setItems()`.

I would say, that the old `beginUpdate(false)` mode is just a bug in the whole system, and correct behavior is `beginUpdate(true)`, so it should be only that variant after all. I'm right?

**Comment by 6pac** [comment]
Perhaps in practice, but the regular ```suspend``` only suppresses refresh. That's it. To add suppression of a bunch of other thungs would be a breaking change. All I did was prevent the Unique Id check occurring in bulk update mode and then call it in the ```endUpdate()```
